dds/DCPS/DataReaderImpl.cpp

```diff
@@ -320,7 +320,12 @@ DataReaderImpl::add_association(const RepoId& yourId,
   //is held here anyway
   guard.release();
 
-  if (!associate(data, active)) {
+  if (associate(data, active)) {
+    Observer::Rch observer = get_observer(Observer::e_ASSOCIATED);
```

Comment:
```suggestion
    const Observer::Rch observer = get_observer(Observer::e_ASSOCIATED);
```
There are many of these so I haven't marked each one.  Please add `const` to local variables that aren't changed after initialization.

Reply:
I strongly agree that a variable (not only local) unchanged after initialization should be defined as const. This also entails that only const methods of that object are accessible. In this particular case, we can make const Observer::Rch observer only if we also make all the three groups of interface methods const. That was what I was not sure about in the design phase, i.e. how these interface methods and most importantly their derived implementations will be used. For example, in the Observer test, TestObserver needs to update a specific statistics when a specific event handler is called. It will be difficult to take a specific action like that if these interface methods are const. In other words, we want to make our code robust, but we don't want to limit its usefulness.

Reply:
You can add const here without changing other code. It's the `RcHandle` that's being constrained, not the thing being pointed to.  This is the same way that `shared_ptr` works.  There are some examples in the code already.
```
TransportClient.cpp:677:  const DataLink_rch link = found->second;
```

Reply:
I see. 
It's very confusing that after this line: const DataLink_rch link = found->second;
the link is then used to call these mutating methods: 
link->release_reservations(peerId, repo_id_, released);
link->remove_listener(repo_id_);

These behaviors are made possible by RcHandle's const operator->, that is: 
T* operator->() const
{
    return this->ptr_;
}

Technically, this changes the meaning of the const keyword in variable definition here, if not completely defeats its purpose.

Changes are made as suggested.

Reply:
It is working as designed, and just as shared_ptr does.  This is the difference between `const T*` and `T* const`.